### PR TITLE
[API] enable paggination for ListRepoTags

### LINKED ISSUE #10453: API for List a repository's tags
The api currently does not provide paging function. If there is a large amount of data, there will be problems. Can you add a paging mechanism?

image

### PR DESCRIPTION
**Pull request by 6543**
close #10453

### CODE REVIEW

File: modules/git/repo_tag.go [hunk 1]
```diff
@@ -187,14 +187,27 @@ func (repo *Repository) GetTag(name string) (*Tag, error) {
 }
 
 // GetTagInfos returns all tag infos of the repository.
-func (repo *Repository) GetTagInfos() ([]*Tag, error) {
+func (repo *Repository) GetTagInfos(page, pageSize int) ([]*Tag, error) {
```
Comment (6543) [inline modules/git/repo_tag.go:190]: I can not use models.ListOptions because **import cycle**

File: modules/git/repo_tag.go [hunk 2]
```diff
@@ -187,14 +187,27 @@ func (repo *Repository) GetTag(name string) (*Tag, error) {
...
 	// TODO this a slow implementation, makes one git command per tag
 	stdout, err := NewCommand("tag").RunInDir(repo.Path)
```
Comment (6543) [inline modules/git/repo_tag.go:192]: I didn't found a good solution to limmit outcome of tag directly via git so I'll filter afterwards

**Review by lafriks — APPROVED** [review]

File: modules/git/repo_tag.go [hunk 3]
```diff
@@ -187,14 +187,27 @@ func (repo *Repository) GetTag(name string) (*Tag, error) {
...
+	if page != 0 {
+		skip := (page - 1) * pageSize
+		if skip >= len(tagNames) {
+			return nil, nil
+		}
+		tagNames = tagNames[skip:]
+		if len(tagNames) < pageSize {
+			pageSize = len(tagNames)
+		}
+		tagNames = tagNames[:pageSize]
```
Comment (lafriks) [inline modules/git/repo_tag.go]: You could calculate length so that you could use: `tagNames = tagNames[skip:skip+pageSize]`

Comment (6543) [reply modules/git/repo_tag.go]: @lafriks this produce weared outcome: https://play.golang.org/p/n5OCMLM1U-O

File: routers/api/v1/repo/tag.go [hunk 4]
```diff
@@ -30,11 +31,21 @@ func ListTags(ctx *context.APIContext) {
 	//   description: name of the repo
 	//   type: string
 	//   required: true
+	// - name: page
+	//   in: query
+	//   description: page number of results to return (1-based)
+	//   type: integer
+	// - name: limit
+	//   in: query
+	//   description: page size of results, maximum page size is 50
```
Comment (guillep2k) [inline routers/api/v1/repo/tag.go]:
```suggestion
	//   description: page size of results, default maximum page size is 50
```

File: templates/swagger/v1_json.tmpl [hunk 5]
```diff
@@ -7251,6 +7251,18 @@
...
+          },
+          {
+            "type": "integer",
+            "description": "page number of results to return (1-based)",
+            "name": "page",
+            "in": "query"
+          },
+          {
+            "type": "integer",
+            "description": "page size of results, maximum page size is 50",
```
Comment (guillep2k) [inline templates/swagger/v1_json.tmpl]:
```suggestion
            "description": "page size of results, default maximum page size is 50",
```

**Review by guillep2k — APPROVED** [review]
Just a couple of nits. Maximum page size is actually `setting.API.MaxResponseItems` which defaults to 50.

**Comment by techknowlogick** [comment]
please "update branch"

**Comment by 6543** [comment]
@techknowlogick done ;)

**Pull request merged**